plugins/telink-dfu/fu-telink-dfu-hid-device.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright 2024 Mike Chang <Mike.chang@telink-semi.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "config.h"
+
+#include <linux/hidraw.h>
+#include <linux/input.h>
+
+#include "fu-telink-dfu-archive.h"
+#include "fu-telink-dfu-hid-device.h"
+#include "fu-telink-dfu-struct.h"
+
+struct _FuTelinkDfuHidDevice {
+	FuHidDevice parent_instance;
+	;
+};
+
+G_DEFINE_TYPE(FuTelinkDfuHidDevice, fu_telink_dfu_hid_device, FU_TYPE_HID_DEVICE)
+
+#define FU_TELINK_DFU_HID_DEVICE_START_ADDR	0x0000
+#define FU_TELINK_DFU_HID_DEVICE_REPORT_TIMEOUT 500 /* ms */
+#define FU_TELINK_DFU_HID_DEVICE_OTA_LENGTH	64
+#define FU_TELINK_DFU_HID_DEVICE_REPORT_ID	6
+#define HID_IFACE				2
+#define HID_EP_IN				(0x80 | 4)
+#define HID_EP_OUT				(0x00 | 5)
+
+static gboolean
+fu_telink_dfu_hid_device_setup(FuDevice *device, GError **error)
+{
+	/* HidDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_telink_dfu_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static FuStructTelinkDfuHidPkt *
+fu_telink_dfu_hid_device_create_packet(FuTelinkDfuCmd cmd,
+				       const guint8 *buf,
+				       gsize bufsz,
+				       GError **error)
+{
+	FuStructTelinkDfuHidPkt *pkt = fu_struct_telink_dfu_hid_pkt_new();
+	guint16 ota_data_len;
+
+	switch (cmd) {
+	case FU_TELINK_DFU_CMD_OTA_FW_VERSION:
+		ota_data_len = 0;
+		break;
+	case FU_TELINK_DFU_CMD_OTA_START:
+		ota_data_len = 2;
+		break;
+	case FU_TELINK_DFU_CMD_OTA_END:
+		ota_data_len = 6;
+		break;
+	default:
+		/* OTA data */
+		ota_data_len = 20;
+		break;
+	}
+
+	fu_struct_telink_dfu_hid_pkt_set_report_id(pkt, FU_TELINK_DFU_HID_DEVICE_REPORT_ID);
+	fu_struct_telink_dfu_hid_pkt_set_ota_data_len(pkt, ota_data_len);
+	fu_struct_telink_dfu_hid_pkt_set_ota_cmd(pkt, cmd);
+	if (buf != NULL)
+		fu_struct_telink_dfu_hid_pkt_set_payload(pkt, buf, bufsz, NULL);
```

Comment:
I think you need to use error and check the return value here

Reply:
resolved in commit 0fb74d587e8aedf6a6bccabff38623ea034d224d